options-design/mongo/clientopts/clientopts.go

```diff
@@ -0,0 +1,90 @@
+package clientopts
+
+import (
+	"context"
+	"net"
+	"time"
+)
+
+type ContextDialer interface {
+	DialContext(ctx context.Context, network, address string) (net.Conn, error)
+}
+
+type Option interface {
+	option()
+}
+
+type OptionFunc func(*Client) error
+
+type Client struct{}
+
+type Bundle struct{}
+
+func BundleOptions(...Option) Bundle { return Bundle{} }
+
+func (b *Bundle) Appname(string) *Bundle                                  { return nil }
+func (b *Bundle) AuthMechanism(string) *Bundle                            { return nil }
+func (b *Bundle) AuthMechaismProperties(map[string]string) *Bundle        { return nil }
+func (b *Bundle) AuthSource(s string) *Bundle                             { return nil }
+func (b *Bundle) ConnectTimeout(d time.Duration) *Bundle                  { return nil }
+func (b *Bundle) Dialer(d ContextDialer) *Bundle                          { return nil }
+func (b *Bundle) HeartbeatInterval(d time.Duration) *Bundle               { return nil }
+func (b *Bundle) Hosts(s []string) *Bundle                                { return nil }
+func (b *Bundle) Journal(v bool) *Bundle                                  { return nil }
+func (b *Bundle) LocalThreshold(d time.Duration) *Bundle                  { return nil }
+func (b *Bundle) MaxConnIdleTime(d time.Duration) *Bundle                 { return nil }
+func (b *Bundle) MaxConnsPerHost(u uint16) *Bundle                        { return nil }
+func (b *Bundle) MaxIdleConnsPerHost(u uint16) *Bundle                    { return nil }
+func (b *Bundle) Password(s string) *Bundle                               { return nil }
+func (b *Bundle) ReadConcernLevel(s string) *Bundle                       { return nil }
+func (b *Bundle) ReadPreference(s string) *Bundle                         { return nil }
+func (b *Bundle) ReadPreferenceTagSets(m []map[string]string) *Bundle     { return nil }
+func (b *Bundle) MaxStaleness(d time.Duration) *Bundle                    { return nil }
+func (b *Bundle) ReplicaSet(s string) *Bundle                             { return nil }
+func (b *Bundle) ServerSelectionTimeout(d time.Duration) *Bundle          { return nil }
+func (b *Bundle) Single(v bool) *Bundle                                   { return nil }
+func (b *Bundle) SocketTimeout(d time.Duration) *Bundle                   { return nil }
+func (b *Bundle) SSL(v bool) *Bundle                                      { return nil }
+func (b *Bundle) SSLClientCertificateKeyFile(s string) *Bundle            { return nil }
+func (b *Bundle) SSLClientCertificateKeyPassword(s func() string) *Bundle { return nil }
+func (b *Bundle) SSLInsecure(v bool) *Bundle                              { return nil }
+func (b *Bundle) SSLCaFile(s string) *Bundle                              { return nil }
+func (b *Bundle) WString(s string) *Bundle                                { return nil }
```

Comment:
Can they just provide a WriteConcern object?

Reply:
Same as ReadPreference.  There are several options that have to be accounted for, e.g. wTimeout.

Reply:
There are options for each. I'm not sure if we want to figure out a way to combine WString and WNumber, or if a user would ever want to put a WTimeout in the connection string and then pass a WString or WNumber as an option. The argument here is the same as that for read preference.